File: examples/python/test_farthest_point_sampling.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import numpy as np
import time
import open3d as o3d
import cv2
import torch

import misc3d as m3d
from IPython import embed


""" numpy implementation of farthest point sampling """
def farthest_point_sampling_numpy(xyz, npoint):
    N = xyz.shape[0]
    indices = [0] * npoint
    distance = np.ones((N, )) * 1e10
    farthest = 0
    for i in range(npoint):
        indices[i] = farthest
        centroid = xyz[farthest, :]
        dist = np.sum((xyz - centroid)**2, -1)
        mask = dist < distance
        distance[mask] = dist[mask]
        farthest = np.argmax(distance)
    return indices


pcd = o3d.io.read_point_cloud('../data/pose_estimation/model/obj.ply')
print('before smapling: {}'.format(pcd))

points = np.asarray(pcd.points)

t0 = time.time()
indices = m3d.preprocessing.farthest_point_sampling(pcd, 1000)
print('time cost for misc3d: {}'.format(time.time() - t0))
sample = pcd.select_by_index(indices)

t0 = time.time()
indices = farthest_point_sampling_numpy(points, 1000)
print('time cost for numpy: {}'.format(time.time() - t0))
sample_numpy = pcd.select_by_index(indices)

vis = o3d.visualization.Visualizer()
vis.create_window("Farest point sampling", 1920, 1200)
m3d.vis.draw_point_cloud(vis, pcd)
m3d.vis.draw_point_cloud(vis, sample, color=(0, 1, 0), size=5)
vis.run()

File: examples/cpp/test_farthest_point_sampling.cpp
#include <iostream>
#include <memory>

#include <misc3d/preprocessing/filter.h>
#include <misc3d/utils.h>
#include <misc3d/vis/vis_utils.h>
#include <open3d/geometry/PointCloud.h>
#include <open3d/io/PointCloudIO.h>
#include <open3d/visualization/visualizer/Visualizer.h>

int main(int argc, char *argv[]) {
    open3d::geometry::PointCloud pcd;
    bool ret = open3d::io::ReadPointCloud(
        "../examples/data/pose_estimation/model/obj.ply", pcd);

    std::cout << "Point cloud size before smapling: " << pcd.points_.size()
              << std::endl;

    misc3d::Timer timer;
    timer.Start();
    auto indices = misc3d::preprocessing::FarthestPointSampling(pcd, 1000);
    std::cout << "Time cost: " << timer.Stop() << std::endl;

    auto sample = pcd.SelectByIndex(indices);

    auto vis = std::make_shared<open3d::visualization::Visualizer>();
    vis->CreateVisualizerWindow("Farthest point sampling", 1920, 1200);
    misc3d::vis::DrawPointCloud(vis, pcd);
    misc3d::vis::DrawPointCloud(vis, *sample, {0, 1, 0},
                                Eigen::Matrix4d::Identity(), 5);
    vis->Run();

    return 0;
}